### PR TITLE
Trollop Option Parsing, Per-Side Margins

### PR DESCRIPTION
**Pull request by kemitchell**
Added support for per-side margins:

`wkpdf --margins 10.0`

or

`wkpdf --margins 10.0 15.0`

or equivalently

`wkpdf --margins 10.0 15.0 10.0 15.0`

Also ported the command line parser to Trollop http://trollop.rubyforge.org/, which provides some nice features (short flags, data type enforcement, etc.).

Thanks for developing wkpdf! I use it daily to format legal opinions, which are nice to print with one wide margin for notes.

Cheers,

KEM
Austin, Texas

### CODE REVIEW

**Comment by plessl** [comment]
Hi Kem

Basically I like the added functionality and I could also live with moving to the Trollop option parser. What is bothering me however is `wkpdf --help` is not presenting the user with as much information on the available options and possible settings as previously. For example, my version showed the user what paper sizes are valid. It this information is added in a proper way, I will be happy to integrate your patch into mainline.

Christian

**Comment by kemitchell** [comment]
Great. I'm on it.

**Comment by kemitchell** [comment]
kemitchell/wkpdf@50a83bd07c563fb92729's `--help` screen appears to be showing a pipe-separated list of all valid paper specifiers. What else did you have in mind? I've pushed commits that show valid options when the user enters an invalid one. Was there anything else?

**Comment by plessl** [comment]
Great, I like it. What was your rationale to include the trollop source rather than requiring the installation of the trollop gem? Just convenience?

**Comment by kemitchell** [comment]
I've added trollop 1.16.2 or newer to the `Rakefile` gemspec dependencies list and removed `trollop.rb`.

I originally added the whole source file because other than RubyCocoa there weren't any dependencies. If you don't mind `require 'rubygems'`, I agree that's the “right” way to go.

**Comment by plessl** [comment]
Thanks, the patch looks good. I plan to merge at the end of next week.

File: Rakefile [hunk 1]
```diff
@@ -16,6 +16,7 @@ begin
     gemspec.email        = "wkpdf@plesslweb.ch"
     gemspec.homepage     = "http://plessl.github.com/wkpdf"
     gemspec.authors      = ["Christian Plessl"]
+	gemspec.add_dependency "trollop", ">= 1.16.2"
```
Comment (kemitchell) [inline Rakefile:19]: This was just the newest on my system. I didn't check to see if older versions might work as well.

File: lib/controller.rb [hunk 2]
```diff
@@ -157,11 +155,11 @@ class Controller < NSObject
     p = CommandlineParser.instance
     viewToPrint = webView.mainFrame.frameView.documentView
     r = viewToPrint.bounds
-    if p.margin > 0 then
-      r.origin.x -= p.margin;
-      r.origin.y -= p.margin;
-      r.size.width += 2 * p.margin;
-      r.size.height += 2 * p.margin;
+    if p.margins.any?{|x| x > 0} then
```
Comment (kemitchell) [inline lib/controller.rb:158]: This is where the idiocy started. My Cocoa is pretty rusty (and the different origins always through me when I wasn't), so I basically tested this empirically. Hopefully it's logically sound.

**Comment by plessl** [comment]
After some testing, I have now merged all of your changes. I applied some minor changes, see a495434a529aa7372c43b898a7a1d09fb5664f1c and e37ba9fb3b9e9c7db3df03a00aa6528d6d0332a9.

The new version of wkpdf is now available from RubyGems as version 0.4.0. Since the command line arguments have been changed a bit I thought it is worth a bump of the minor revision.

Thanks for your contribution.

**Comment by kemitchell** [comment]
Thanks for wkpdf!

**Pull request merged**